<a href="https://colab.research.google.com/github/KarmaticNeutral/cse380-notebooks/blob/master/W07/07_2_Ponder_and_Prove_Elementary_Number_Theory.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove Elementary Number Theory
#### Due: Saturday, 20 February 2021, 11:59 pm.

## Explore Fermat's Little Theorem Further


Fermat's Little Theorem (FLT) says that if $N$ is prime, then $N$ divides $X^N - X$.

Remember, the contrapositive of the conditional statement in this theorem is that if $N$ **doesn't** divide $X^N - X$ for some $X$, then $N$ **can't** be prime.

Unfortunately, this simple primality test doesn't always work, because it can be fooled by **pseudoprimes**.

For example, $341 = 11 \cdot 31$ is not prime. But $341$ **does** divide $2^{341} - 2$ as verified below:

In [ ]:
((2 ** 341) - 2) % 341

0

So $341$ is a so-called **base-2 pseudoprime**. What about **base-3**?

In [ ]:
((3 ** 341) - 3) % 341

165

Check that the result is not zero, therefore $341$ is **not** a **base-3 pseudoprime**.

Are there any other bases that will not fool the FLT test for $341$?

Are there any pseudoprimes that will fool the FLT for **any choice** of base coprime to them?

### The answer is yes.

Your task is the find the first 4-digit **bullet-proof pseudoprime** (**bppp**) and **prove** (yes, **PROVE**) that it will fool the FLT test for every base coprime to it.

Your proof must use all of the following:
1. the definition of coprime,
2. a consequence of coprimality,
3. the factorization of the **bppp**,
4. FLT, and the
5. CRT (Chinese Remainder Theorem).


In [ ]:
from math import gcd
from sympy import isprime

def passes_FLT_test_even_though_not_prime(b, n):
  # primes don't count as pseudoprimes
  return not isprime(n) and (b ** n) % n == b

def is_bppp(n):
  bases_coprime_to_n = [b for b in range(2, n) if gcd(b, n) == 1]
  return all(list(map(lambda b: passes_FLT_test_even_though_not_prime(b, n), 
                      bases_coprime_to_n)))

n = 1000
while not is_bppp(n):
  n += 1

n

1105

###Proof

Fermat's Little Therom: $X^{N - 1} \equiv_N 1$

1. We know that 1105 is composite since it is the product of 5, 13 and 17.

2. We know that two numbers are coprime of to each other if their GCD is 1.

3. It is a consequence of coprimality that if a and b are comprime, then b is also coprime to the prime factors of a. In our case: gcd(1105, b) = gcd(5, b) = 1.

4. We then apply Fermat's Little Therom:

  $b^{5-1} = b^{4} \equiv_5 1$

  $b^{13-1} = b^{12} \equiv_{13} 1$

  $b^{17-1} = b^{16} \equiv_{17} 1$

5. Since we know that $b^{1104} \equiv_1105
1$, we know that the following are also true:
  
  $(b^4)^{276} = b^{1104} \equiv_{5} 1$
  
  $(b^{12})^{92} = b^{1104} \equiv_{13} 1$
  
  $(b^{16})^{69} = b^{1104} \equiv_{17} 1$

6. By the Chinese Remainder Therom:

  $x \equiv_p r$ and $x \equiv_q r$ and $x \equiv_s r$, then $x \equiv_pqs r$

7. Therefore 1105 foes fool Fermat's Little Therom for all possible 'b's coprime to it.

8. QED

$\color{red}{\text{Excellent work!}}$

# DONE My Report on What I Did and What I Learned

## Fun
This was a difficult assignment for me to get into. Once I did however, I was able to quickly begin enjoying my time. I love these kinds of assignments, the manipulation of numbers, as long as I feel I have some idea of direction I should be going.

## New
As I haven't really worked with proofs before, this was a difficult task. It required me to consider the same problems I have been doing from a different perspective. For example, I have often worked with GCDs and primality in the past, it just took me some time to start looking at those same tools in this new light.

## Meaningful
My answer to this question is very similar to that of the previous. I explored proofs and became far more familiar with them. This is a part of logic that is not going to leave my life any time soon. Therefore, becoming more familar is significant and meaningful to me.

## Other
Due to how timing worked out and the fact that I was very sick on Saturday and wasn't able to complete the assignment that day. I used many video and refreshers on Fermat's Little Theorom, number theroy, primality and more to help me work through this problem.

# DONE What is True?
Click on each warranted checkbox to toggle it to True (or back to False). 

NOTE: *This only works in Colab. If you run it in some other Jupyter notebook client/server environment you may have to change False to True (or vice versa) manually.*

This self-assessment is subject to revision by a grader.

In [ ]:
#@markdown ## What is True about what I did?
#@markdown ### I had fun.
cb00 = True #@param {type:'boolean'}
#@markdown ### I learned something new.
cb01 = True #@param {type:'boolean'}
#@markdown ### I achieved something meaningful, or something I can build upon at a later time.
cb02 = True #@param {type:'boolean'}
#@markdown ## What is true about my report?
#@markdown ### I wrote a sufficient number of well-written sentences.
cb03 = True #@param {type:'boolean'}
#@markdown ### My report is free of mechanical infelicities.
cb04 = True #@param {type:'boolean'}
#@markdown ### I used Grammarly (or something better described in my report) to check for MIs.
cb05 = True #@param {type:'boolean'}
#@markdown ### I reported on any connections I found between these problems and something I already know. 
cb06 = True #@param {type:'boolean'}
#@markdown ### I reported who were and what contribution each of my collaborators made.
cb07 = False #@param {type:'boolean'}
#@markdown ## What is true about my proof?
#@markdown ### It succinctly uses the definition of coprime.
cb08 = True #@param {type:'boolean'}
#@markdown ### It correctly uses the definition of coprime.
cb09 = True #@param {type:'boolean'}
#@markdown ### It succinctly uses a consequence of comprimality
cb10 = True #@param {type:'boolean'}
#@markdown ### It correctly uses a consequence of comprimality
cb11 = True #@param {type:'boolean'}
#@markdown ### It succinctly uses the factorization of the **bppp**,
cb12 = True #@param {type:'boolean'}
#@markdown ### It correctly uses the factorization of the **bppp**,
cb13 = True #@param {type:'boolean'}
#@markdown ### It succinctly uses Fermat's Little Theorem. 
cb14 = True #@param {type:'boolean'}
#@markdown ### It correctly uses Fermat's Little Theorem. 
cb15 = True #@param {type:'boolean'}
#@markdown ### It succinctly uses the Chinese Remainder Theorem.
cb16 = True #@param {type:'boolean'}
#@markdown ### It correctly uses the Chinese Remainder Theorem.
cb17 = True #@param {type:'boolean'}



$\color{red}{\text{Forgot to mention who you worked/didn't work with!}}$